default.nix

```diff
@@ -18,6 +18,32 @@ let
   ]
   ++ programs.modules;
 
+  # treefmt-nix can be loaded into a submodule. In this case we get our `pkgs` from
+  # our own standard option `pkgs`; not externally.
+  submodule-modules = [
+    ({ config, lib, pkgs, ... }:
+      let
+        inherit (lib)
+          mkOption
+          types;
+      in
+      {
+        options.pkgs = mkOption {
+          type = types.uniq (types.lazyAttrsOf (types.raw or types.unspecified));
+          description = ''
+            Nixpkgs to use in `treefmt`.
+          '';
+          default = pkgs;
+          defaultText = "`pkgs` (module argument)";
```

Comment:
Could you move these two back to the flake-parts module?
That will fix the accidental circular definition of `pkgs`, and you can also make the default text more descriptive:
```nix
defaultText = "`pkgs` (module argument of `perSystem`)";
```

Option _declarations_ are also merged, as long as the option attribute don't conflict.